endpoint/endpoint_test.go

```diff
@@ -216,3 +216,136 @@ func TestIsOwnedBy(t *testing.T) {
 		})
 	}
 }
+
+func TestDuplicatedEndpointsWithSimpleZone(t *testing.T) {
+	foo1 := &Endpoint{
+		DNSName:    "foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "foo",
+		},
+	}
+	foo2 := &Endpoint{
+		DNSName:    "foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "foo",
+		},
+	}
+	bar := &Endpoint{
+		DNSName:    "foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "bar",
+		},
+	}
+
+	type args struct {
+		eps []*Endpoint
+	}
+	tests := []struct {
+		name string
+		args args
+		want []*Endpoint
+	}{
+		{
+			name: "filter values",
+			args: args{
+				eps: []*Endpoint{
+					foo1,
+					foo2,
+					bar,
+				},
+			},
+			want: []*Endpoint{
+				foo1,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := RemoveDuplicates(tt.args.eps); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("RemoveDuplicates() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestDuplicatedEndpointsWithOverlappingZones(t *testing.T) {
+	foo1 := &Endpoint{
+		DNSName:    "internal.foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "foo",
+		},
+	}
+	foo2 := &Endpoint{
+		DNSName:    "internal.foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "foo",
+		},
+	}
+	foo3 := &Endpoint{
+		DNSName:    "foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "foo",
+		},
+	}
+	foo4 := &Endpoint{
+		DNSName:    "foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "foo",
+		},
+	}
+	bar := &Endpoint{
+		DNSName:    "internal.foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "bar",
+		},
+	}
+	bar2 := &Endpoint{
+		DNSName:    "foo.com",
+		RecordType: RecordTypeA,
+		Labels: Labels{
+			OwnerLabelKey: "bar",
+		},
+	}
+
+	type args struct {
+		eps []*Endpoint
+	}
+	tests := []struct {
+		name string
+		args args
+		want []*Endpoint
+	}{
+		{
+			name: "filter values",
+			args: args{
+				eps: []*Endpoint{
+					foo1,
+					foo2,
+					foo3,
+					foo4,
+					bar,
+					bar2,
+				},
+			},
+			want: []*Endpoint{
```

Comment:
So with this implementation, the first instance of an duplicate endpoint is the one that's kept. Is that an agreeable implementation detail? Thinking out loud here... `registry.Records()` is implemented by each provider, e.g. AWS, which calls (a variant of) `ListResourceRecordSets` under the hood which "sorts results first by DNS name with the labels reversed... When multiple records have the same DNS name, ListResourceRecordSets sorts results by the record type."

ref: https://pkg.go.dev/github.com/aws/aws-sdk-go/service/route53#hdr-Sort_order

If yes, then this PR seems a good solution to me.

Reply:
The concern is that even using the EndpointKey function, global and weighted records don't have enough test coverage and I'm not familiar if they use the same targets, but that's what dedupsource is using to generate the primary key, without the recordtype. The real questions here are "What use cases do we have that needs endpoints with the same "Key" but different targets?" "What should the endpoint's primary key be? Does it need to change across providers?"